AlgebraInLean/World0LeanIntroSOLUTIONS/Sheet5Solutions.lean

```diff
@@ -0,0 +1,159 @@
+
+Let's begin with "simp". "Simp" uses lemmas already in mathlib as well as existing hypotheses to
+simplify the goal. "Simp" has a wide range of uses, so if you are curious about uses beyond the very
+simple examples shown below, looking up the documentation is recommended. For example, just as
+"exact" and "apply" have "exact?" and "apply?", "simp" has variation "simp?".
+
+Let's look at a couple examples:-/
+
+/-NNG Inequality World Level 1-/
+example (x : Nat) : x ≤ x := by
+  simp
+  done
+
+example (x y : Nat) (h1 : x = 13) (h2 : y = 2) : x + y = 15 := by
+  simp [h2]
+  exact h1
+  done
+
+/-Now, see if you can complete the following exercises.-/
+
+/-NNG Inequality World Level 2-/
+example (x : Nat) : 0 ≤ x := by
+  simp
+  done
+
+example (x : Nat) : x = x := by
+  simp
+  done
+
+example (x y : Nat) (h1 : x = 13) (h2 : y = 2) : x + y = 15 := by
+  simp [h1, h2]
+  done
+
+/-Moving on to the "use" tactic. A quick shortcut that indicates whether this tactic will be handy
+is if "∃" appears. "Use" will replace that variable with whatever you have input.
+
+For example:-/
+
+example : ∃ x : Nat, x + 3 = 34 := by
+  use 31
+  done
+
+/-Complete the exercises below. "Simp" could also come in handy here.-/
+
+example : ∃ x y : Nat, 5 * x + 3 * y = 13 := by
+  use 2
+  use 1
+  done
+
+example : ∃ x y z : Nat, 4 * y + 30 / z - 21 * x = 5 := by
+  use 1, 5, 5
+  done
+
+example : ∃ x z : ℤ, x * z = y := by
+  use y
+  use 1
+  simp
+  done
+
+/-The "symm" tactic is used to change a goal such as a ∼ b to b ∼ a. It works only when the relation
+is symmetric or has previously been proven to be symmetric.
+
+For example:-/
+
+example (x y : Nat) (h1 : x = y) : y = x := by
+  symm
+  exact h1
+  done
+
+/-Here's a quick exercise for you:-/
+
+example (x y z : Nat) (h1 : x = y * z) (h2 : z = 4) : y * 4 = x := by
+  rw [h2] at h1
+  symm
+  exact h1
+  done
+
+/-The "induction" tactic is similar to the "cases" tactic in that it uses a specific structure, and
+VSCode can automatically generate the structure once you type something like "induction x" (or
+"induction x with"). Note that this is different from how you used induction in the Natural Number
+Game, just as the cases tactic is different.
+
+Try out these exercises using induction:-/
+
+/-NNG Advanced Addition World Level 4-/
+example (x y : Nat) : x + y = x → y = 0 := by
+  induction x with
+  | zero =>
+    intro h1
+    rw [zero_add] at h1
+    exact h1
+  | succ n ih =>
+    intro h2
+    simp at h2
+    exact h2
+  done
+
+/-NNG Advanced Addition World Level 5-/
+example (a b : Nat) : a + b = 0 → a = 0 := by
+  induction a with
+  | zero =>
+    intro h1
+    rfl
+  | succ n ih =>
+    intro h2
+    simp [h2] at ih
+    simp [ih] at h2
+    done
+
+/-The "have" tactic should also be familiar to you. With the "have" tactic, the user can add a new
+hypothesis to the list of existing hypotheses after proving it is true.
```

Comment:
`have` is an important tactic, as it lends itself to the use of "forwards logic", in that you build up a hypothesis to match the goal. This is one often used in proofs on paper. However, Lean proofs tend to be more elegant if they use "backwards logic", as in changing the goal to match a hypothesis (for example, the "apply" tactic uses backwards logic by default). I think this is something important for a new user to consider, though we should emphasize that both approaches are valid.

Reply:
Note added.